pymc3/dp/dp.py

```diff
@@ -0,0 +1,297 @@
+#   Copyright 2021 The PyMC Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+import sys
+
+import warnings
+
+sys.path.insert(0, "../../../pymc3")
+
+import numpy as np
+import pandas as pd
+
+from aesara import tensor as at
+
+import pymc3 as pm
+
+import aesara
+from aesara.tensor.random.op import RandomVariable
+
+aesara.config.exception_verbosity = "high"
+
+from pymc3.math import logsumexp
+
+import argparse
+
+__all__ = [
+    "DirichletProcess",
+]
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--dp", action="store_true")
+parser.add_argument("--dpmix", action="store_true")
+args = parser.parse_args()
+
+
+def stick_breaking(betas):
+    R"""
+    betas: vector of Beta random variables
+    """
+    # need to make sure that betas is one-dimensional
+
+    # if (betas.shape.eval()[0] <= 2):
+    #     raise AttributeError(
+    #         "More betas are needed to generate stick-breaking weights."
+    #     )
+
+    sticks = at.concatenate(
+        [
+            [1],
+            at.cumprod(1 - betas[:-1]),
+        ]
+    )
+
+    return betas * sticks
```

Comment:
Does `betas * sticks` be equal to 1?

The thing I could not understand is -
Lets say I set K=4, then lets represent betas as `b1, b2, b3, b4`.
With the code snippet in this function - 

```
betas * sticks = b1 * 1 + b2 * (1-b1) + b3 * (1-b1) * (1-b2) + b4 * (1-b1) * (1-b2) * (1-b3)
```

Aren't we missing `(1-b1) * (1-b2) * (1-b3) * (1-b4)` to make it one?

Reply:
You are correct! In the case of K = 4, we should effectively append (1 - b1) * ... * (1 - b4) to the list of weights we wish to output